ext/openssl/openssl.c

```diff
@@ -6625,63 +6639,56 @@ static int php_openssl_cipher_update(const EVP_CIPHER *cipher_type,
 }
 /* }}} */
 
-/* {{{ proto string openssl_encrypt(string data, string method, string password [, int options=0 [, string $iv=''[, string &$tag = ''[, string $aad = ''[, int $tag_length = 16]]]]])
-   Encrypts given data with given method and key, returns raw or base64 encoded string */
-PHP_FUNCTION(openssl_encrypt)
+
+PHP_OPENSSL_API zend_string* php_openssl_encrypt(char *data, size_t data_len, char *method, size_t method_len, char *password, size_t password_len, zend_long options, char *iv, size_t iv_len, zval *tag, zend_long tag_len, char *aad, size_t aad_len)
 {
-	zend_long options = 0, tag_len = 16;
-	char *data, *method, *password, *iv = "", *aad = "";
-	size_t data_len, method_len, password_len, iv_len = 0, aad_len = 0;
-	zval *tag = NULL;
 	const EVP_CIPHER *cipher_type;
 	EVP_CIPHER_CTX *cipher_ctx;
 	struct php_openssl_cipher_mode mode;
-	int i=0, outlen;
-	zend_string *outbuf;
+	int i = 0, outlen;
 	zend_bool free_iv = 0, free_password = 0;
+	zend_string *outbuf = NULL;
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "sss|lszsl", &data, &data_len, &method, &method_len,
-					&password, &password_len, &options, &iv, &iv_len, &tag, &aad, &aad_len, &tag_len) == FAILURE) {
-		return;
-	}
+	PHP_OPENSSL_CHECK_SIZE_T_TO_INT_NORET(data_len, data);
+	PHP_OPENSSL_CHECK_SIZE_T_TO_INT_NORET(password_len, password);
+	PHP_OPENSSL_CHECK_SIZE_T_TO_INT_NORET(aad_len, aad);
+	PHP_OPENSSL_CHECK_LONG_TO_INT_NORET(tag_len, tag_len);
 
-	PHP_OPENSSL_CHECK_SIZE_T_TO_INT(data_len, data);
-	PHP_OPENSSL_CHECK_SIZE_T_TO_INT(password_len, password);
-	PHP_OPENSSL_CHECK_SIZE_T_TO_INT(aad_len, aad);
-	PHP_OPENSSL_CHECK_LONG_TO_INT(tag_len, tag_len);
 
 	cipher_type = EVP_get_cipherbyname(method);
 	if (!cipher_type) {
 		php_error_docref(NULL, E_WARNING, "Unknown cipher algorithm");
-		RETURN_FALSE;
+		return NULL;
 	}
 
 	cipher_ctx = EVP_CIPHER_CTX_new();
 	if (!cipher_ctx) {
 		php_error_docref(NULL, E_WARNING, "Failed to create cipher context");
-		RETURN_FALSE;
+		return NULL;
 	}
+
 	php_openssl_load_cipher_mode(&mode, cipher_type);
 
 	if (php_openssl_cipher_init(cipher_type, cipher_ctx, &mode,
 				&password, &password_len, &free_password,
 				&iv, &iv_len, &free_iv, NULL, tag_len, options, 1) == FAILURE ||
 			php_openssl_cipher_update(cipher_type, cipher_ctx, &mode, &outbuf, &outlen,
 				data, data_len, aad, aad_len, 1) == FAILURE) {
-		RETVAL_FALSE;
+		if (outbuf)
+			zend_string_release_ex(outbuf, 0);
+		outbuf = 0;
```

Comment:
This should probably be `outbuf = NULL`? Same in a few cases below.

I'm also wondering why this code needed a change. Not that it's necessarily wrong, but I would expect that if outbuf needs to be freed here, then that have also been the case before this refactoring. Or am I missing something?

Reply:
By itself the change is not feature changing for sure but it gives the possiblity to uses the api outside of this module (I have further plans personally).

Reply:
Sorry, I didn't mean the overall change, but the added zend_string_release call here.

Reply:
When I launched the unit tests I got leak warnings. Not against changing to NULL for the rest.